Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp

```diff
@@ -0,0 +1,373 @@
+    source_link_hash_t&& sourceLinkHash,
+    source_link_equality_t&& sourceLinkEquality) const {
+  auto measurementIndexMap =
+      std::unordered_map<SourceLink, std::size_t, source_link_hash_t,
+                         source_link_equality_t>(0, sourceLinkHash,
+                                                 sourceLinkEquality);
+
+  std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
+      measurementsPerTrack;
+
+  ACTS_INFO("Starting to compute initial state");
+
+  for (const auto& track : tracks) {
+    std::vector<std::tuple<std::size_t, std::size_t, bool>> measurements_tuples;
+
+    for (auto ts : track.trackStatesReversed()) {
+      if (ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        Acts::SourceLink sourceLink = ts.getUncalibratedSourceLink();
+
+        const auto& geoID = ts.referenceSurface().geometryId();
+
+        // assign a new measurement index if the source link was not seen yet
+        auto emplace = measurementIndexMap.try_emplace(
+            sourceLink, measurementIndexMap.size());
+
+        bool isoutliner =
+            ts.typeFlags().test(Acts::TrackStateFlag::OutlierFlag);
+
+        measurements_tuples.push_back(
+            std::make_tuple(emplace.first->second, geoID.volume(), isoutliner));
+      }
+    }
+
+    measurementsPerTrack.push_back(std::move(measurements_tuples));
+  }
+
+  return measurementsPerTrack;
+}
+
+template <typename track_container_t, typename traj_t,
+          template <typename> class holder_t>
+std::vector<double> Acts::AthenaAmbiguityResolution::simpleScore(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    std::vector<std::map<std::size_t, Counter>>& counterMaps,
+    Optional_cuts<track_container_t, traj_t, holder_t> optionalCuts) const {
+  std::vector<double> trackScore;
+
+  int iTrack = 0;
+
+  ACTS_INFO("Number of detectors: " << m_cfg.detectorMap.size());
+
+  ACTS_INFO("Starting to score tracks");
+
+  // Loop over all the trajectories in the events
+  for (const auto& track : tracks) {
+    auto counterMap = std::map<std::size_t, Counter>();
+
+    for (const auto& ts : track.trackStatesReversed()) {
+      auto iVolume = ts.referenceSurface().geometryId().volume();
+
+      ACTS_DEBUG("Volume: " << iVolume);
+
+      auto iTypeFlags = ts.typeFlags();
+
+      auto volume_it = m_cfg.volumeMap.find(iVolume);
+      if (true) {
+        auto detectorId = volume_it->second;
+
+        if (iTypeFlags.test(Acts::TrackStateFlag::HoleFlag)) {
+          counterMap[detectorId].nHoles++;
+        } else if (iTypeFlags.test(Acts::TrackStateFlag::MeasurementFlag)) {
+          if (iTypeFlags.test(Acts::TrackStateFlag::SharedHitFlag)) {
+            counterMap[detectorId].nSharedHits++;
+          }
+          counterMap[detectorId].nHits++;
+        } else if (iTypeFlags.test(Acts::TrackStateFlag::OutlierFlag)) {
+          counterMap[detectorId].nOutliers++;
+        }
+      } else {
```

Comment:
You should use the same if ordering as in the MultiTrajectoryHelper (meas first) as it is faster

Reply:
It was like that at first, but i changed it for double holes, I will rectify it soon.